content/blog/options/index.md

```diff
@@ -0,0 +1,237 @@
+             <std::result::Result<T, F> as FromResidual<std::result::Result<Infallible, E>>>
+```
+
+([Rust Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=7001ff6af6c0bcbf44249691d65b086f))
+
+Ouch. This is scary-looking.
+There's a lot of visual noise in this error message. The `FromResidual` and `Yeet` are implementation details that are not relevant to the user,
+but the relevant details are somewhat obscured. 
+We're not making a good first impression here.
+
+**And all we did was use an `Option`!**
+
+My main gripe with this error message is that it doesn't explain **why** the `?` operator doesn't work with `Option`...
+just that it doesn't.
+Granted, the error message is already quite long and it might not be the best place to teach Rust, but it would be nice to have a link to a more detailed explanation. After all, this is Rust: we take great pride in our error messages.
+
+By the way, running `rustc --explain E0277` only gives you a very generic explanation about types which don't implement traits. 
+That's not very actionable.
+
+## What Do People Usually Do?
+
+The most common solution I see is that people are confused for a bit,
+try to understand the error message, and eventually just give up, use `unwrap`,
+and make a mental note to come back to it later.
+
+```rust
+fn get_user_name() -> Result<String, String> {
+    let user = get_user().unwrap();
+    // Do something with `user`
+    Ok(user)
+}
+```
+
+Even with a trainer around, they are often too embarrassed to ask for help.
+They think people are supposed to "get this" and they are the only ones who don't.
+
+This just defers the problem. 
+The user of the function will get a panic at runtime. It might be your future self.
+
+Moreover, I advise people to avoid `unwrap` in production code, but this sets a bad example.
+
+During trainings, this is where I pause for a moment and explain the error message.
+It's a great point where we dig deeper to understand the root cause of the problem.
+
+Okay, I've kept you waiting long enough. Let's demystify this error message.
+
+## Why doesn't `?` work with `Option`?
+
+The `?` operator takes a `Result` and unwraps it, returning the value inside if it is `Ok`, or returning early if it is `Err`. 
+
+Essentially, `?` is implemented like this:
+
+```rust
+let value = match my_result {
+    Ok(value) => value,
+    Err(error) => return Err(error),
+};
+```
+
+`Option` on the other hand **does not have an** `Err` variant. 
+
+It is defined like this:
+
+```rust
+enum Option<T> {
+    Some(T),
+    None,
+}
+```
+
+`None` doesn't have an associated error value. It's just a value that means "no value".
+The `?` operator wouldn't know what error it should return:
+
+```rust
+let value = match my_option {
+    Some(value) => value,
+    None => todo!("What to return here? ¯\_(ツ)_/¯"),
+};
+```
+
+That's actually the entire difference between `Result` and `Option`.
```

Comment:
That's not true. There is also a semantic difference. You go into some detail later in this paragraph, but I don't think you can just state this blanket statement. Maybe qualify it with "the entire implementation difference"? Or "technical difference"?

Reply:
Clarified it. Thanks.